File: forensics/oracle/probes/probe_tables.c
/*
 * probe_tables.c — oracle probe for the rcode/tsigrcode/type/class tables
 * (the TABLES-0001 court).
 *
 * Reads one command per line from stdin: `<cmd> <arg>`.  Emits one result
 * line per command:
 *
 *   OK <value>
 *   ERR <result-text>
 *
 * Commands:
 *   rcode-totext <n>        dns_rcode_totext
 *   rcode-fromtext <tok>    dns_rcode_fromtext (value or error text)
 *   tsigrcode-totext <n>    dns_tsigrcode_totext
 *   tsigrcode-fromtext <tok>
 *   type-totext <n>         dns_rdatatype_totext
 *   type-fromtext <tok>     dns_rdatatype_fromtext
 *   type-ismeta <n>         dns_rdatatype_ismeta -> 0/1
 *   type-issingleton <n>    dns_rdatatype_issingleton -> 0/1
 *   type-isknown <n>        dns_rdatatype_isknown -> 0/1
 *   class-totext <n>        dns_rdataclass_totext
 *   class-fromtext <tok>    dns_rdataclass_fromtext
 *
 * Oracle-side tooling only (spec §2).
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include <isc/buffer.h>
#include <isc/mem.h>
#include <isc/region.h>
#include <isc/result.h>
#include <isc/util.h>

#include <dns/rcode.h>
#include <dns/rdataclass.h>
#include <dns/rdatatype.h>
#include <dns/result.h>

/* Not in the public header (lib/dns/rdata.c 9.20.26). */
extern bool dns_rdatatype_ismeta(dns_rdatatype_t type);
extern bool dns_rdatatype_issingleton(dns_rdatatype_t type);
extern bool dns_rdatatype_isknown(dns_rdatatype_t type);

static isc_mem_t *mctx = NULL;

static void
print_totext(isc_result_t result, const char *buf, unsigned int len) {
	if (result != ISC_R_SUCCESS) {
		printf("ERR %s\n", isc_result_totext(result));
	} else {
		printf("OK %.*s\n", (int)len, buf);
	}
}

int
main(void) {
	char line[512];

	isc_mem_create(&mctx);

	while (fgets(line, sizeof(line), stdin) != NULL) {
		char cmd[64], arg[256];
		unsigned long n;
		isc_textregion_t tr;
		isc_buffer_t b;
		char buf[512];
		isc_result_t result;

		if (sscanf(line, "%63s %255s", cmd, arg) != 2) {
			if (sscanf(line, "%63s", cmd) != 1) {
				continue;
			}
			arg[0] = '\0';
		}

		tr.base = (unsigned char *)arg;
		tr.length = (unsigned int)strlen(arg);

		if (strcmp(cmd, "rcode-totext") == 0 ||
		    strcmp(cmd, "tsigrcode-totext") == 0 ||
		    strcmp(cmd, "type-totext") == 0 ||
		    strcmp(cmd, "class-totext") == 0 ||
		    strcmp(cmd, "type-ismeta") == 0 ||
		    strcmp(cmd, "type-issingleton") == 0 ||
		    strcmp(cmd, "type-isknown") == 0)
		{
			n = strtoul(arg, NULL, 10);
			isc_buffer_init(&b, buf, sizeof(buf));
			if (strcmp(cmd, "rcode-totext") == 0) {
				result = dns_rcode_totext((dns_rcode_t)n, &b);
				print_totext(result, buf, isc_buffer_usedlength(&b));
			} else if (strcmp(cmd, "tsigrcode-totext") == 0) {
				result = dns_tsigrcode_totext((dns_rcode_t)n, &b);
				print_totext(result, buf, isc_buffer_usedlength(&b));
			} else if (strcmp(cmd, "type-totext") == 0) {
				result = dns_rdatatype_totext((dns_rdatatype_t)n, &b);
				print_totext(result, buf, isc_buffer_usedlength(&b));
			} else if (strcmp(cmd, "class-totext") == 0) {
				result = dns_rdataclass_totext((dns_rdataclass_t)n, &b);
				print_totext(result, buf, isc_buffer_usedlength(&b));
			} else if (strcmp(cmd, "type-ismeta") == 0) {
				printf("OK %d\n", dns_rdatatype_ismeta((dns_rdatatype_t)n) ? 1 : 0);
			} else if (strcmp(cmd, "type-issingleton") == 0) {
				printf("OK %d\n", dns_rdatatype_issingleton((dns_rdatatype_t)n) ? 1 : 0);
			} else {
				printf("OK %d\n", dns_rdatatype_isknown((dns_rdatatype_t)n) ? 1 : 0);
			}
			continue;
		}

		if (strcmp(cmd, "rcode-fromtext") == 0 ||
		    strcmp(cmd, "tsigrcode-fromtext") == 0 ||
		    strcmp(cmd, "type-fromtext") == 0 ||
		    strcmp(cmd, "class-fromtext") == 0)
		{
			unsigned int value = 0;
			if (strcmp(cmd, "rcode-fromtext") == 0) {
				result = dns_rcode_fromtext((dns_rcode_t *)&value, &tr);
			} else if (strcmp(cmd, "tsigrcode-fromtext") == 0) {
				result = dns_tsigrcode_fromtext((dns_rcode_t *)&value, &tr);
			} else if (strcmp(cmd, "type-fromtext") == 0) {
				result = dns_rdatatype_fromtext((dns_rdatatype_t *)&value, &tr);
			} else {
				result = dns_rdataclass_fromtext((dns_rdataclass_t *)&value, &tr);
			}
			if (result != ISC_R_SUCCESS) {
				printf("ERR %s\n", isc_result_totext(result));
			} else {
				printf("OK %u\n", value);
			}
			continue;
		}

		printf("ERR unknown-command\n");
	}

	isc_mem_destroy(&mctx);
	return 0;
}

File: bind9-rs-forensics/src/bin/probe-tables.rs
//! `probe-tables` — bind9-rs side of the `TABLES-0001` court.
//! Mirrors `forensics/oracle/probes/probe_tables.c`: each stdin line is
//! `<cmd> <arg>`; one result line per command:
//!
//! ```text
//! OK <value>
//! ERR <result-text>
//! ```
//!
//! The commands cover `dns_rcode_totext/fromtext`,
//! `dns_tsigrcode_totext/fromtext`, `dns_rdatatype_totext/fromtext` and the
//! ismeta/issingleton/isknown predicates, and
//! `dns_rdataclass_totext/fromtext` — the RCODE/class/type tables.

use bind9_rs_core::class::Class;
use bind9_rs_core::error::Error;
use bind9_rs_core::rcode::{tsigrcode_from_text, tsigrcode_to_text, Rcode};
use bind9_rs_core::rrtype::RrType;
use std::io::{BufRead, Write};

fn err_text(e: &Error) -> String {
    e.bind_totext().to_string()
}

fn main() {
    let stdin = std::io::stdin();
    let stdout = std::io::stdout();
    let mut out = stdout.lock();
    for line in stdin.lock().lines() {
        let line = line.unwrap_or_default();
        let line = line.trim_end_matches(['\n', '\r']);
        if line.is_empty() {
            continue;
        }
        let mut parts = line.splitn(2, ' ');
        let Some(cmd) = parts.next() else { continue };
        // The C probe reads the argument with %s, so only the first
        // whitespace-delimited word is the token.
        let arg = parts
            .next()
            .unwrap_or("")
            .split_whitespace()
            .next()
            .unwrap_or("");
        let res = match cmd {
            "rcode-totext" => Ok(Rcode::from_u16(parse_n(arg)).to_text()),
            "tsigrcode-totext" => Ok(tsigrcode_to_text(parse_n(arg))),
            "type-totext" => Ok(RrType::from_u16(parse_n(arg)).to_text()),
            "class-totext" => Ok(Class::from_u16(parse_n(arg)).to_text()),
            "type-ismeta" => Ok(if RrType::from_u16(parse_n(arg)).is_meta() {
                "1".to_string()
            } else {
                "0".to_string()
            }),
            "type-issingleton" => Ok(if RrType::from_u16(parse_n(arg)).is_singleton() {
                "1".to_string()
            } else {
                "0".to_string()
            }),
            "type-isknown" => Ok(if RrType::from_u16(parse_n(arg)).is_known() {
                "1".to_string()
            } else {
                "0".to_string()
            }),
            "rcode-fromtext" => Rcode::from_text(arg).map(|r| r.to_u16().to_string()),
            "tsigrcode-fromtext" => tsigrcode_from_text(arg).map(|n| n.to_string()),
            "type-fromtext" => RrType::from_text(arg).map(|t| t.to_u16().to_string()),
            "class-fromtext" => Class::from_text(arg).map(|c| c.to_u16().to_string()),
            _ => Err(Error::Other("unknown-command".to_string())),
        };
        match res {
            Ok(v) => {
                let _ = writeln!(out, "OK {v}");
            }
            Err(e) => {
                let _ = writeln!(out, "ERR {}", err_text(&e));
            }
        }
    }
}

fn parse_n(s: &str) -> u16 {
    s.parse().unwrap_or(0)
}
